### PR TITLE
[SPARK-30050][SQL] analyze table and rename table should not erase hive table bucketing info

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch adds Hive provider into table metadata in `HiveExternalCatalog.alterTableStats`. When we call `HiveClient.alterTable`, `alterTable` will erase if it can not find hive provider in given table metadata.

Rename table also has this issue.

### Why are the changes needed?

Because running `ANALYZE TABLE` on a Hive table, if the table has bucketing info, will erase existing bucket info.

### Does this PR introduce any user-facing change?

Yes. After this PR, running `ANALYZE TABLE` on Hive table, won't erase existing bucketing info.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by viirya** [comment]
cc @cloud-fan

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogSuite.scala [hunk 1]
```diff
@@ -108,6 +108,46 @@ class HiveExternalCatalogSuite extends ExternalCatalogSuite {
...
+    assert(bucketString1.contains("10"))
+
+    catalog.alterTableStats("db1", "t", None)
+
+    val bucketString2 = externalCatalog.client.runSqlHive("DESC FORMATTED db1.t")
+      .filter(_.contains("Num Buckets")).head
+    assert(bucketString2.contains("10"))
+  }
+
+  test("SPARK-30050: rename table should not erase the bucketing metadata at hive side") {
```
Comment (viirya) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogSuite.scala]: I think I can combine two tests.

**Comment by viirya** [comment]
cc @felixcheung

**Comment by viirya** [comment]
retest this please.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala [hunk 2]
```diff
@@ -537,7 +537,17 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
       rawTable.storage
     }
 
-    val newTable = rawTable.copy(
+    // Fills in Hive provider if necessary. Otherwise, `Hive.alterTable` will overwrite
+    // important table metadata like bucketing when altering table.
+    val tableMeta = rawTable.properties.get(DATASOURCE_PROVIDER) match {
+      // This is a Hive serde table, set its provider.
+      case None if rawTable.tableType != VIEW =>
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: Does it mean this table is created by Hive?

Comment (viirya) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: This just means the table is Hive serde table. I think we can also use Spark command to create Hive table.

**Comment by cloud-fan** [comment]
Shall we update `HiveClientImpl.toHiveTable`? We shouldn't use `DDLUtils.isHiveTable(table)` to check if it's a hive table

**Comment by viirya** [comment]
> Shall we update HiveClientImpl.toHiveTable? We shouldn't use DDLUtils.isHiveTable(table) to check if it's a hive table. We should use !DDLUtils.isDatasourceTable

In this case, because we get raw table metadata by getRawTable, the provider info is not propagated yet.

We should use HiveExternalCatalog.isDatasourceTable, which will also check provider in table properties.

**Comment by cloud-fan** [comment]
thanks, merging to master/2.4!

**Comment by viirya** [comment]
thanks!

**Pull request closed without merging**